In [1]:
import pandas as pd

df = pd.read_csv('../../../etl//raw_data/교육부 국사편찬위원회_한국역사용어시소러스 정보_20211028 (1).csv')

df.info()

<class 'pandas.DataFrame'>
RangeIndex: 62409 entries, 0 to 62408
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   term_id         62409 non-null  int64
 1   topterm_id      62409 non-null  int64
 2   term_name       62409 non-null  str  
 3   term_kind       62409 non-null  int64
 4   term_ch         61852 non-null  str  
 5   term_remark     12889 non-null  str  
 6   term_attr       358 non-null    str  
 7   term_year       61356 non-null  str  
 8   term_times      60543 non-null  str  
 9   term_lk         61903 non-null  str  
 10  term_desc       61583 non-null  str  
 11  term_user       62409 non-null  int64
 12  term_created    62409 non-null  str  
 13  term_reference  34101 non-null  str  
dtypes: int64(4), str(10)
memory usage: 6.7 MB


In [2]:
df.tail(2)

,term_id,topterm_id,term_name,term_kind,term_ch,term_remark,term_attr,term_year,term_times,term_lk,term_desc,term_user,term_created,term_reference
62407,4910,8,사악,2,司樂,NaN,NaN,759-776,남북국시대,정치·행정·법제>인사,신라시대 국가의 음악을 맡은 대악감(大樂監)의 장관.,10,2003-10-07 19:40:21,한국민족문화대백과사전
62408,4912,8,사알,2,司謁,고려,NaN,?-?,고려후기,정치·행정·법제>인사,고려후기 내시부(內侍府)의 정7품 환관직.,10,2003-08-24 17:56:45,NaN


In [3]:
df['term_kind'].value_counts()

term_kind
2    61598
1      794
0       17
Name: count, dtype: int64

In [5]:
df[df['term_kind'] == 0].head(2)

,term_id,topterm_id,term_name,term_kind,term_ch,term_remark,term_attr,term_year,term_times,term_lk,term_desc,term_user,term_created,term_reference
2954,1,1,역사일반,0,,,NaN,,NaN,_NULL_,,0,2002-09-30 22:42:00,NaN
2961,8,8,정치·행정·법제,0,,,NaN,,NaN,_NULL_,,0,2002-09-30 22:42:00,NaN


In [6]:
df[df['topterm_id'] == 662].head(2)

,term_id,topterm_id,term_name,term_kind,term_ch,term_remark,term_attr,term_year,term_times,term_lk,term_desc,term_user,term_created,term_reference
61,8855,662,혜왕,2,惠王,NaN,NaN,?-599,삼국시대,인명,백제 제28대 왕. 재위 598-599년. 제27대 위덕왕(威德王)의 동생.,18,2003-04-23 12:58:29,NaN
544,21027913,662,타추간,2,陁鄒干,NaN,NaN,?,삼국시대,인명,신라초기 주변에 있던 음즙벌국(音汁伐國)의 왕. 102년(파사이사금 23) 8월 신...,10,2005-04-11 11:49:04,삼국사기


In [23]:
df['topterm_id'].value_counts()

topterm_id
8      9355
662    9024
665    5352
664    4889
439    4884
333    4207
631    3403
518    3387
199    2945
49     2921
407    2502
86     2406
663    1869
574    1658
1      1499
283    1086
310    1022
Name: count, dtype: int64

In [7]:
df[df['term_kind'] == 1].tail(2)


,term_id,topterm_id,term_name,term_kind,term_ch,term_remark,term_attr,term_year,term_times,term_lk,term_desc,term_user,term_created,term_reference
61933,48239148,86,대한제국기,1,大韓帝國期,일본외교,NaN,NaN,NaN,NaN,NaN,23,2004-02-23 11:27:41,NaN
61934,48239151,86,일제시기,1,日帝時期,일본외교,NaN,NaN,NaN,NaN,NaN,23,2004-02-23 11:27:41,NaN


In [8]:
df['term_remark'].value_counts()

term_remark
        7538
고려       455
조선       386
문화재      171
인명       145
        ... 
상설아문       1
신라관계       1
진흥왕        1
배삯         1
사회생활       1
Name: count, Length: 1720, dtype: int64

In [9]:
df['term_lk'].value_counts()

term_lk
인명                                    8623
문화재                                   5181
서명                                    4738
정치·행정·법제>인사                           4650
정치·행정·법제>행정>중앙행정기구                    1915
                                      ... 
문화·예술>종합예술>>문화·예술>음악                     1
사회·생활>풍속·의례>풍속>>사회·생활>일상생활>의생활           1
정치·행정·법제>사법>사법제도>>정치·행정·법제>행정>행정제도       1
인명>성격별                                   1
교통·통신>교통시설>>교통·통신>교통로                    1
Name: count, Length: 383, dtype: int64

In [10]:
df['term_user'].value_counts()

term_user
0     14753
10     9268
8      8211
7      3637
6      3519
5      3148
14     2703
3      2666
17     2230
4      1933
19     1921
23     1717
16     1700
18     1678
11     1142
2       684
9       625
12      437
1       242
21       87
25       56
13       28
20       12
22       11
15        1
Name: count, dtype: int64

In [16]:
df[df['term_user'] == 8 ].head(2)

,term_id,topterm_id,term_name,term_kind,term_ch,term_remark,term_attr,term_year,term_times,term_lk,term_desc,term_user,term_created,term_reference
5,5626,439,수궁가,2,水宮歌,NaN,NaN,?-?,조선후기,문화·예술>음악,조선후기에 불리는 판소리 다섯마당 가운데 하나.,8,2004-07-19 16:00:16,한국민족문화대백과사전
16,5641,333,수렵총,2,狩獵塚,NaN,NaN,?-?,삼국시대,사회·생활>풍속·의례>매장,고구려의 벽화고분으로 평안남도 남포시 와우도구역 화도리에 있음.,8,2002-10-23 03:55:47,한국민족문화대백과사전


## EDA 정리: `term_kind`, `topterm_id`, `term_lk` 구조 해석

현재 확인한 기준으로 `term_kind`는 행의 역할을 나타내는 값으로 본다.

- `term_kind = 0`: 원본 최상위 대분류 카테고리
- `term_kind = 1`: 하위 분류/태그/색인어 후보
- `term_kind = 2`: 실제 역사 용어 본문 항목

`topterm_id`는 직접 부모 ID라기보다, 해당 행이 속한 최상위 대분류의 `term_id`로 보인다. `topterm_id.value_counts()` 결과가 17개이고, 이 값들이 `term_kind = 0`인 대분류의 `term_id`와 일치하기 때문이다.

예를 들면 `정치·행정·법제`는 `term_kind = 0`, `term_id = 8`, `topterm_id = 8`이고, 그 아래의 하위 분류나 실제 용어는 `topterm_id = 8`을 공유한다. 따라서 `topterm_id`만으로는 `범죄` 같은 하위 분류 안에 어떤 실제 용어가 들어가는지 알 수 없다.

실제 용어의 상세 분류 경로는 `term_lk`에 들어 있다. 예시는 다음과 같다.

```text
정치·행정·법제>인사
정치·행정·법제>행정>중앙행정기구
사회·생활>풍속·의례>풍속>>사회·생활>일상생활>의생활
```

`>`는 계층 구분자로 보고, `>>`는 한 용어가 여러 분류 경로를 가질 때의 복수 경로 구분자로 보는 것이 자연스럽다.

따라서 1차 전처리에서는 다음 흐름이 적절하다.

1. `term_kind = 0`으로 원본 대분류 17개를 확보한다.
2. `term_lk`를 `>>` 기준으로 먼저 분리해 복수 분류 경로를 나눈다.
3. 각 분류 경로를 `>` 기준으로 다시 분해한다.
4. 분해된 경로 조각으로 `Category` 후보를 만든다.
5. 실제 용어인 `term_kind = 2` 행을 해당 `Category`에 연결한다.
6. `term_kind = 1`은 직접 부모 관계로 확정하지 않고, `term_lk`에서 만든 카테고리 후보를 검증하거나 보강하는 참고 데이터로 사용한다.

`범죄` 하위 분류의 실제 용어를 확인하려면 `topterm_id`만 보지 말고 `term_lk` 또는 `term_remark`를 같이 확인해야 한다.

```python
df[
    (df["term_kind"] == 2) &
    (df["topterm_id"] == 8) &
    (df["term_lk"].fillna("").str.contains("범죄"))
][["term_id", "term_name", "topterm_id", "term_lk", "term_remark", "term_desc"]]
```


## EDA 정리: 1차 전처리 컬럼 사용 기준

현재까지 확인한 기준으로 `history_terms.csv`는 `Term`, `Category`, `TopCategory`, `Period` 후보를 뽑는 원천 데이터로 사용한다. 핵심은 실제 용어 행인 `term_kind = 2`를 중심으로 보고, 분류 체계는 `term_lk`를 `>>`와 `>`로 분해해서 만드는 것이다.

### 1. 컬럼 기반 추출 가능 데이터

| 뽑을 데이터 | 주로 쓰는 컬럼 |
|---|---|
| `Term` 노드 | `term_id`, `term_name`, `term_ch`, `term_desc` |
| 원본 대분류 `TopCategory` | `term_kind=0`, `term_id`, `topterm_id`, `term_name` |
| 상세 `Category` | `term_lk`를 `>>`, `>`로 분해 |
| `Term - HAS_CATEGORY - Category` 관계 | `term_id`, `term_lk` |
| 시대/기간 후보 `Period` | `term_times`, `term_year` |
| `start_year`, `end_year`, `date_precision` | `term_year` |
| `Term - IN_PERIOD / INFERRED_IN_PERIOD - Period` | `term_times`, `term_year`, 나중에 `term_desc` |
| 엔티티 타입 후보 | `topterm_id`, `term_lk`, `term_name`, `term_desc` |
| 주제/키워드 후보 | `term_lk`, `term_desc` |
| 중복/별칭 검수 후보 | `term_name`, `term_ch`, `term_remark`, `term_desc` |
| 장소 후보 | `topterm_id=663`, `term_lk=지명`, `term_desc` |
| 서명/문헌 후보 | `topterm_id=664`, `term_lk=서명` |
| 문화재 후보 | `topterm_id=665`, `term_lk=문화재` |
| 관직/기관 후보 | `term_lk`의 `인사`, `중앙행정기구`, `기관`, `term_desc` |
| 검수 후보 | 중복 이름, 복수 `term_lk`, 불명확한 `term_year`, 애매한 `term_remark` |

### 2. Neo4j 1차 MVP에 사용할 컬럼

| 원본 컬럼 | 사용 방향 | 설명 |
|---|---|---|
| `term_id` | 사용 | `Term` 노드의 고유 ID로 사용한다. `term_name` 중복이 많으므로 ID 기준이 필요하다. |
| `topterm_id` | 사용 | 원본 최상위 대분류 연결 기준이다. `term_kind = 0`의 `term_id`와 연결된다. |
| `term_name` | 사용 | `Term.name`으로 사용한다. 단, 고유키로 쓰지 않는다. |
| `term_kind` | 사용 | 행의 역할 구분값이다. `0`은 대분류, `1`은 하위 분류/태그 후보, `2`는 실제 용어다. |
| `term_ch` | 사용 | 한자/원문 표기이므로 `Term.hanja` 속성으로 넣는다. |
| `term_remark` | 보조 사용 | 동명이인, 세부 구분, 검수 단서로 사용한다. 핵심 관계 기준은 아니다. |
| `term_year` | 사용 | `year_text`로 보존하고, 이후 `start_year`, `end_year`, `date_precision` 파싱에 사용한다. |
| `term_times` | 사용 | `period_text`로 보존하고, `Period` 후보 및 `Term - IN_PERIOD` 후보 생성에 사용한다. |
| `term_lk` | 핵심 사용 | 카테고리 경로다. `>>`는 복수 경로, `>`는 계층 구분자로 사용한다. |
| `term_desc` | 사용 | 설명문이다. 검색/RAG, 주제 추론, 장소/기관/인물 후보 추출에 사용한다. |

### 3. 1차 MVP에서 제외하거나 원본 메타로만 둘 컬럼

| 원본 컬럼 | 판단 | 이유 |
|---|---|---|
| `term_attr` | 1차 제외 | 결측이 많고 현재 의미가 불명확하다. |
| `term_user` | 제외 | 원본 시스템의 입력자/관리자 ID로 보인다. 역사 지식 의미와 무관하다. |
| `term_created` | 제외 또는 원본 메타 보존 | 역사적 시점이 아니라 원본 DB 등록 시각이다. |
| `term_reference` | 선택 보존 | 출처 메타데이터다. 1차 카테고리/용어 그래프에는 필수는 아니지만 나중에 `Source` 노드나 답변 근거에 쓸 수 있다. |

### 4. `term_lk` 기반 카테고리 세분화 원칙

```text
>> : 복수 카테고리 경로 구분자
>  : 단일 경로 안의 계층 구분자
```

예시:

```text
사회·생활>풍속·의례>풍속>>사회·생활>일상생활>의생활
```

분해 결과:

```text
1번 경로: 사회·생활 > 풍속·의례 > 풍속
2번 경로: 사회·생활 > 일상생활 > 의생활
```

따라서 `Term`은 각 경로의 마지막 `Category`와 `HAS_CATEGORY`로 연결하고, `Category`끼리는 `SUBCATEGORY_OF`로 연결한다.

### 5. 지금 단계에서 만들 수 있는 산출물

- `normalized_terms.csv`: 실제 용어 행 정규화 결과
- `category_dictionary.csv`: `term_lk`를 분해해서 만든 카테고리 노드 후보
- `term_category_relation.csv`: `Term`과 최하위 `Category` 연결
- `period_dictionary.csv` 후보: `term_times`, `term_year` 기반 시대 후보
- `term_period_relation.csv` 후보: `Term`과 `Period` 연결
- `review_candidates.csv`: 중복 이름, 복수 카테고리, 불명확한 연도, 애매한 `term_remark` 검수 후보

### 6. 1차 Neo4j MVP 구조

노드:

```text
Term
Category
TopCategory
Period
```

관계:

```text
Term - HAS_CATEGORY -> Category
Category - SUBCATEGORY_OF -> Category 또는 TopCategory
Term - HAS_TOP_CATEGORY -> TopCategory
Term - IN_PERIOD 또는 INFERRED_IN_PERIOD -> Period
```

현재 EDA 결론상 다음 작업은 `term_lk` 분해 로직을 먼저 만들고, 그 결과로 `category_dictionary.csv`와 `term_category_relation.csv`를 만드는 것이다.
